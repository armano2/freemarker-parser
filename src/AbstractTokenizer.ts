export default abstract class AbstractTokenizer {
  protected template : string = ''
  protected length : number = 0
  protected index : number = 0

  protected charAt (i : number) : string {
    return this.template.charAt(i)
  }

  protected charCodeAt (i : number) : number {
    return this.template.charCodeAt(i)
  }

  protected init (template : string) : void {
    this.template = template
    this.length = template.length
    this.index = 0
  }
}
