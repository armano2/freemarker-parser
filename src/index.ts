import { Parser } from './Parser'

export {
  Parser,
}
