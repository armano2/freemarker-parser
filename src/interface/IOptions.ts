export interface IOptions {
  squareTags?: boolean;
  // squareInterpolation? : boolean // TODO: unsupported
  // hashInterpolation? : boolean // TODO: unsupported
  parseLocation?: boolean;
  // useCamelCaseNaming? : boolean // TODO: unsupported
}
