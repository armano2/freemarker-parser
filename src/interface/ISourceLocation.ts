export interface ISourceLocation {
  line: number;
  column: number;
}
