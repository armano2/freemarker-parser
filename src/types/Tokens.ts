import NodeNames from '../enum/NodeNames'
import { ENodeType } from '../Symbols'

export interface IDirectivesTypes {
  [n : string] : NodeNames
}

export interface IToken extends ILoc {
  type : ENodeType
  startTag? : string
  endTag? : string
  params? : string
  text : string
  isClose : boolean
}

export interface ILoc {
  start : number
  end : number
}
