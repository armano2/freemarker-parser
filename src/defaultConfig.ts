import { Options } from './interface/Options';

const defaultConfig: Options = {
  squareTags: false,
  parseLocation: true,
};

export default defaultConfig;
