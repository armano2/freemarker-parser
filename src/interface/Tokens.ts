import { ENodeType } from '../Symbols'
import { ISourceLocation } from './ISourceLocation'

export interface ILoc {
  start : number
  end : number,
  loc? : {
    start : ISourceLocation,
    end : ISourceLocation,
  }
}

export interface IToken extends ILoc {
  type : ENodeType
  startTag? : string
  endTag? : string
  params? : string
  text : string
}
