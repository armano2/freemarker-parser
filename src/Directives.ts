import NodeNames from './enum/NodeNames'
import { IDirectivesTypes } from './types/Tokens'

const Directives : IDirectivesTypes = {
  if: NodeNames.Condition,
  else: NodeNames.Else,
  elseif: NodeNames.ConditionElse,
  list: NodeNames.List,
  include: NodeNames.Include,
  assign: NodeNames.Assign,
  attempt: NodeNames.Attempt,
  compress: NodeNames.Compress,
  // escape: NodeNames.Escape,
  // noescape: NodeNames.NoEscape,
  // fallback: NodeNames.Fallback,
  function: NodeNames.Function,
  // flush: NodeNames.Flush,
  global: NodeNames.Global,
  import: NodeNames.Import,
  local: NodeNames.Local,
  // lt: NodeNames.Lt,
  macro: NodeNames.Macro,
  // nested: NodeNames.Nested,
  // nt: NodeNames.Nt,
  recover: NodeNames.Recover,
  // recurse: NodeNames.Recurse,
  return: NodeNames.Return,
  // rt: NodeNames.Rt,
  // setting: NodeNames.Setting,
  // stop: NodeNames.Stop,
  switch: NodeNames.Switch,
  case: NodeNames.SwitchCase,
  default: NodeNames.SwitchDefault,
  break: NodeNames.Break,
  // t: NodeNames.T,
  // visit: NodeNames.Visit,
  noparse: NodeNames.Text,
  noParse: NodeNames.Text,
}

export default Directives
