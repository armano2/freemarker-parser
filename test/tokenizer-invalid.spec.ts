import { Tokenizer } from '../src';
import { IToken } from '../src/interface/Tokens';

const tokenizer = new Tokenizer();

function parse(text: string): IToken[] {
  try {
    return tokenizer.parse(text);
  } catch (e) {
    throw new Error(`${e.message} [${e.start}-${e.end}]`);
  }
}

describe('errors', () => {
  it('invalid amount of brackets', () => {
    expect(() => {
      parse('<#foo foo)>');
    }).toThrowErrorMatchingSnapshot();
    expect(() => {
      parse('<#foo foo(>');
    }).toThrowErrorMatchingSnapshot();
  });

  it('unclosed comment', () => {
    expect(() => {
      parse('<#-- foo bar');
    }).toThrowErrorMatchingSnapshot();
  });

  it('missing close tag in directive', () => {
    expect(() => {
      parse('<#');
    }).toThrowErrorMatchingSnapshot();
  });

  it('missing close tag in macro', () => {
    expect(() => {
      parse('<@');
    }).toThrowErrorMatchingSnapshot();
  });
  it('invalid character in macro name', () => {
    expect(() => {
      parse('<@?');
    }).toThrowErrorMatchingSnapshot();
  });
  it('invalid character in directive name', () => {
    expect(() => {
      parse('<@&');
    }).toThrowErrorMatchingSnapshot();
  });
});

describe('error_expresion', () => {
  it('to many close tags ]', () => {
    expect(() => {
      parse('<@foo a["foo"]]');
    }).toThrowErrorMatchingSnapshot();
  });

  it('to many close tags )', () => {
    expect(() => {
      parse('<@foo a("foo"))');
    }).toThrowErrorMatchingSnapshot();
  });
});

describe('unclosed', () => {
  it('directive', () => {
    expect(() => {
      parse('<#foo');
    }).toThrowErrorMatchingSnapshot();
  });

  it('macro', () => {
    expect(() => {
      parse('<@foo');
    }).toThrowErrorMatchingSnapshot();
  });

  it('interpolation', () => {
    expect(() => {
      parse(`\${ foo`);
    }).toThrowErrorMatchingSnapshot();
  });
});
