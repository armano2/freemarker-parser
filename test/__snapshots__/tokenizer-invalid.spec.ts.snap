// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error_expresion to many close tags ) 1`] = `"To many close tags ) [6-14]"`;

exports[`error_expresion to many close tags ] 1`] = `"To many close tags ] [6-14]"`;

exports[`errors invalid amount of brackets 1`] = `"To many close tags ) [6-9]"`;

exports[`errors invalid amount of brackets 2`] = `"Missing ) close char [6-12]"`;

exports[`errors invalid character in directive name 1`] = `"Invalid \`&\` [2-2]"`;

exports[`errors invalid character in macro name 1`] = `"Invalid \`?\` [2-2]"`;

exports[`errors missing close tag in directive 1`] = `"OpenDirective name cannot be empty [2-2]"`;

exports[`errors missing close tag in macro 1`] = `"OpenMacro name cannot be empty [2-2]"`;

exports[`errors unclosed comment 1`] = `"Unclosed comment [undefined-undefined]"`;

exports[`unclosed directive 1`] = `"Unclosed directive or macro [5-6]"`;

exports[`unclosed interpolation 1`] = `"Unclosed directive or macro [2-7]"`;

exports[`unclosed macro 1`] = `"Unclosed directive or macro [5-6]"`;
